### ISSUE TITLE
“index” vs. function parameter: what’s the difference?

### ISSUE THREAD

**Issue by mathiasbynens** [tc39/proposal-array-unique#3]
Question about <https://github.com/TechQuery/array-unique-proposal#core-features>:

> While `Array.prototype.unique()` invoked with:
> 
> 1.  no parameter, it'll work as `[...new Set(array)]`;
> 
> 2.  one **index-key** parameter (`Number`, `String` or `Symbol`), it'll get values from each array element with the key, and then deduplicates the origin array based on these _values_;
> 
> 3.  one **function** parameter, it'll call this function for each array element, and then deduplicates the origin array based on these _returned values_.

How is 2 different from 3?

**Comment by Jack-Works** [comment]
2 is an abbreviated version of 3.

"key" equals to (x) => x.key

Same for number and symbols.

**Comment by ljharb** [comment]
I think I'd be pretty confused if I learned that shortcut pattern on `unique` and found it wasn't available on all the other predicate-taking array methods.

**Comment by Jack-Works** [comment]
> I think I'd be pretty confused if I learned that shortcut pattern on `unique` and found it wasn't available on all the other predicate-taking array methods.

maybe we can add that

**Comment by ljharb** [comment]
That seems like it'd be better as a separate cross-cutting proposal, if so

**Comment by jridgewell** [comment]
Agreed with @ljharb here. I think we should drop 2 for now, and possibly add it as its own proposal with all of the other array methods (particularly `map`).

**Comment by hax** [comment]
I think unique key is very common use case and I hope it could be included in this proposal.

**Comment by ljharb** [comment]
@hax I agree it's common; but it's already common for every array method that takes a function, not just this new one.

**Comment by hax** [comment]
@ljharb I think it's possible to add key param for other array methods, but consider `unique()` potentially could support multiple keys ( #7 ) which unlikely fit for other methods, I suppose the key usage in `unique` is different than other array methods.

**Comment by ljharb** [comment]
Anything with a predicate seems like it’d apply the same to single or multiple keys.

**Comment by hax** [comment]
Maybe, but I feel `arr.filter(["a", "b"])`, `arr.find("a")`, `arr.some(['a', 'b'])` are confusing.

**Comment by ljharb** [comment]
I do too, but I’d think the same about unique with an array.

**Comment by Jack-Works** [comment]
in the last meeting, the committee requires to remove the `key: string | symbol` signature. it should be a separate proposal to add this kind of feature to all Array methods. @TechQuery

**Comment by TechQuery** [comment]
@ljharb @jridgewell 
Only `map()` may be possible to add a **String-parameter signature**, for example `objects.map('name')`, adding a string parameter to other **Array methods with a callback parameter** may confuse developers:

- `filter()`: keep the elements which has an _Equal string value_ or _Truthy key-value_?
- `findIndex()`: find an element index which has an _Equal string value_ or _Truthy key-value_?
- `some()`: detect an element which has an _Equal string value_ or _Truthy key-value_?
- `every()`: detect every element has an _Equal string value_ or _Truthy key-value_?
- `sort()`: sort elements by a string key, but _Ascending_ or _Descending_?

**Comment by ljharb** [comment]
given the similarity between `.map(() => 'name')` and `.map('name')` i'd say the same potential confusion applies anywhere - or, if not, doesn't apply anywhere.

**Comment by hax** [comment]
I suggest we remove index parameter, and make non-function param throw TypeError so we could reconsider it in the future.
